File: svg/path/cursor.py
class Boolean:
    def __init__(self, val):
        self.val = 1 if (val) else 0

    def __repr__(self):
        return str(self.val)


class Cursor:
    def moveTo(self, x, y):
        self.x = x
        self.y = y
        self.history.append(f"M{x:.3f},{y:.3f}")
        self.edit = True

        return self

    def __init__(self, x, y):
        self.x = 0
        self.y = 0
        self.edit = None
        self.history = []

        self.moveTo(x, y)

    def lineTo(self, x, y):
        if not self.edit:
            return self

        self.x = x
        self.y = y
        self.history.append(f"L{x:.3f},{y:.3f}")

        return self

    def horizontalTo(self, x):
        if not self.edit:
            return self

        self.x = x
        self.history.append(f"H{x:.3f}")

        return self

    def verticalTo(self, y):
        if not self.edit:
            return self

        self.y = y
        self.history.append(f"V{y:.3f}")

        return self

    def quadraticTo(self, cx, cy, x, y):
        if not self.edit:
            return self

        self.x = x
        self.y = y
        self.history.append(f"Q{cx:.3f},{cy:.3f},{x:.3f},{y:.3f}")

        return self

    def cubicTo(self, cx1, cy1, cx2, cy2, x, y):
        if not self.edit:
            return self

        self.x = x
        self.y = y
        self.history.append(f"C{cx1:.3f},{cy1:.3f},{cx2:.3f},{cy2:.3f},{x:.3f},{y:.3f}")

        return self

    def smoothQuadraticTo(self, x, y):
        if not self.edit:
            return self

        self.x = x
        self.y = y
        self.history.append(f"T{x:.3f},{y:.3f}")

        return self

    def smoothCubicTo(self, cx, cy, x, y):
        if not self.edit:
            return self

        self.x = x
        self.y = y
        self.history.append(f"S{cx:.3f},{cy:.3f},{x:.3f},{y:.3f}")

        return self

    def ellipticalArcTo(self, r1, r2, rot, flip, sweep, x, y):
        if not self.edit:
            return self

        self.x = x
        self.y = y
        self.history.append(
            f"A{r1:.3f},{r2:.3f},{rot:.3f},{Boolean(flip)},{Boolean(sweep)},{x:.3f},{y:.3f}"
        )

        return self

    def stopHere(self):
        if not self.edit:
            return self

        self.edit = False
        self.history.append("Z")

        return self

    def toPath(self):
        if not self.edit:
            self.stopHere()

        return " ".join(self.history)

File: lib/svg/elements/path/cursor.cc
#include "cursor.hh"

#include <format>

#include "../utils/concatStr.hh"
#include "../utils/doubleFormat.hh"

Cursor::Cursor(double x, double y)
    : x_{ x }
    , y_{ y }
{
    moveTo(x, y);
}

const std::pair<double, double> Cursor::getPosition() const
{
    return { x_, y_ };
}

void Cursor::moveTo(double x, double y)
{
    x_ = x;
    y_ = y;
    history_.push_back("M" + doubleFmt(x) + "," + doubleFmt(y));

    editing_ = true;
}

void Cursor::lineTo(double x, double y)
{
    if (!editing_)
    {
        return;
    }

    x_ = x;
    y_ = y;
    history_.push_back("L" + doubleFmt(x) + "," + doubleFmt(y));
}

void Cursor::horizontalTo(double x)
{
    if (!editing_)
    {
        return;
    }

    x_ = x;
    history_.push_back("H" + doubleFmt(x));
}

void Cursor::verticalTo(double y)
{
    if (!editing_)
    {
        return;
    }

    y_ = y;
    history_.push_back("V" + doubleFmt(y));
}

void Cursor::quadraticTo(double cx, double cy, double x, double y)
{
    if (!editing_)
    {
        return;
    }

    x_ = x;
    y_ = y;
    history_.push_back("Q" + doubleFmt(cx) + "," + doubleFmt(cy) + ","
                       + doubleFmt(x) + "," + doubleFmt(y));
}

void Cursor::cubicTo(double cx1, double cy1, double cx2, double cy2, double x,
                     double y)
{
    if (!editing_)
    {
        return;
    }

    x_ = x;
    y_ = y;
    history_.push_back("C" + doubleFmt(cx1) + "," + doubleFmt(cy1) + ","
                       + doubleFmt(cx2) + "," + doubleFmt(cy2) + ","
                       + doubleFmt(x) + "," + doubleFmt(y));
}

void Cursor::smoothQuadraticTo(double x, double y)
{
    if (!editing_)
    {
        return;
    }

    x_ = x;
    y_ = y;
    history_.push_back("T" + doubleFmt(x) + "," + doubleFmt(y));
}

void Cursor::smoothCubicTo(double cx, double cy, double x, double y)
{
    if (!editing_)
    {
        return;
    }

    x_ = x;
    y_ = y;
    history_.push_back("S" + doubleFmt(cx) + "," + doubleFmt(cy) + ","
                       + doubleFmt(x) + "," + doubleFmt(y));
}

void Cursor::ellipticalArcTo(double radius1, double radius2, double rot,
                             bool flip, bool sweep, double x, double y)
{
    if (!editing_)
    {
        return;
    }

    x_ = x;
    y_ = y;
    history_.push_back("A" + doubleFmt(radius1) + "," + doubleFmt(radius2) + ","
                       + doubleFmt(rot) + "," + std::to_string(flip) + ","
                       + std::to_string(sweep) + "," + doubleFmt(x) + ","
                       + doubleFmt(y));
}

void Cursor::stopHere()
{
    if (!editing_)
    {
        return;
    }

    editing_ = false;
    history_.push_back("Z");
}

std::string Cursor::toPath()
{
    if (editing_)
    {
        stopHere();
    }

    return joinWithSpace(history_);
}